src/map/zone.cpp

```diff
@@ -692,6 +692,43 @@ void CZone::UpdateWeather()
     // clang-format on
 }
 
+void CZone::CheckMobsPathedBack()
+{
+    // clang-format off
+    CTaskMgr::getInstance()->AddTask("zone_empty_timer", server_clock::now(), this, CTaskMgr::TASK_INTERVAL, 5s,
+    [](time_point tick, CTaskMgr::CTask* PTask)
+    {
+        bool allMobsHomeAndHealed = true;
+        // if the timer runs out, check if all the mobs have pathed home
+        CZone* PZone = std::any_cast<CZone*>(PTask->m_data);
+        if (PZone && PZone->GetZoneEntities())
+        {
+            EntityList_t mobListMap = PZone->GetZoneEntities()->GetMobList();
+            for (const auto& pair : mobListMap)
+            {
+                CMobEntity* mob = dynamic_cast<CMobEntity*>(pair.second);
+
+                // if the mob is not fully healed OR ((the mob can rome home AND is far from home) OR the mob can't rome home)
+                if (mob && !mob->isFullyHealed() || ((mob->CanRoamHome() && mob->IsFarFromHome()) || !mob->CanRoamHome()))
```

Comment:
CI is complaining about this:
```suggestion
                if (mob && (!mob->isFullyHealed() || ((mob->CanRoamHome() && mob->IsFarFromHome()) || !mob->CanRoamHome())))
```

Reply:
i see that ty.  also with print statements im seeing other issues.  looking into them.

Reply:
> Nope, only NMs will path back. Normal mobs will despawn. I guess it could be enabled to work again and this code would help it work properly, which is good

I think this is already supported with a global setting and some mob mods. This logic looks pretty similar to [this in mob_controller](https://github.com/LandSandBoat/server/blob/ce0aad91b7961b9b3d1561110f2a3401803d1c69/src/map/ai/controllers/mob_controller.cpp#L972-L997). Maybe we could add a member var to mob entity, something like `bool m_idle`, and set it in mob_controller on the roam tick, then just check that here.

Reply:
we want to check if the hp and mp has been restored.

which I am running into issues...for example, mobs that are fished up don't report hp==maxhp.  And some random mobs in the zone aren't reporting being fully healed.  Still looking into those.

Reply:
My guess: mobs that aren't spawned are mucking up your check. I bet things would work better if you filtered out the mobs that are not spawned

Reply:
wow yea, that was it.  thank you!

Reply:
> we want to check if the hp and mp has been restored.

The HP check is fine, it's this part I'm looking at:

```cpp
((mob->CanRoamHome() && mob->IsFarFromHome()) || !mob->CanRoamHome())
```
This is basically a duplicate check of this in mob_controller (my comments added):
```cpp
            // PMob->m_idle = true;
            if (!PMob->getMobMod(MOBMOD_DONT_ROAM_HOME) && PMob->IsFarFromHome())
            {
                if (PMob->CanRoamHome())
                {
                    // ...
                    // PMob->m_idle = false;
                }
                else if (!(PMob->getMobMod(MOBMOD_NO_DESPAWN) != 0) && !settings::get<bool>("map.MOB_NO_DESPAWN"))
                {
                    PMob->PAI->Despawn();
                    return;
                }
            }
```
If one of the conditions we're looking for to keep the zone alive is that they're roaming back home and we're already checking for that, we should leverage that instead of duplicating the check. That way you can just check:
```cpp
if (mob && !mob->isFullyHealed() || !mob->m_idle)
```
I'm not sure the exact order this stuff is happening, but it also looks like if this roam check in zone fires before the one in mob_controller, the Despawn call will be delayed until the zone wakes back up.

Reply:
agree.  no need to duplicate code.  I'll change things up and use that.  Thank you!